### DISCUSSION TITLE
Support for additional disks / partitions

### DISCUSSION THREAD

**Discussion by henrygd** [henrygd/beszel#36] [Ideas]
### Update: Initial support has been added in 0.3.0. See [the readme section](https://github.com/henrygd/beszel/blob/main/readme.md#monitoring-additional-disks--partitions) for instructions.

This is a much requested feature (#2, #3, #23) so I want to direct discussion here.

Any opinions on the best way to implement and display would be appreciated. I'd like it to be as straightforward as possible.

Should we just have an environment variable `EXTRA_FILESYSTEMS`  where you would supply a comma separated list?

Then they would be laid out separately like this mock up?

![graphs for usage and i/o under headlines 'root filesystem' and '/dev/sdx'](https://github.com/user-attachments/assets/44480fb8-bd89-483a-89d9-6938f2c2ee0e)

**Comment by bozzio83** [comment]
That sounds great to me!

**Comment by chachoCL** [comment]
🙏

**Comment by crypt0rr** [comment]
Yes, please! 🚀

**Comment by JamsRepos** [comment]
Yeah, good implementation. I saw it as working as all in one array but this makes a little more sense as allows you to monitor the host and arrays seperately. 

Although, perhaps allowing multi file systems for both? Then if people have a pool which is their primary filesystem they can pool those together than the extra for any other system devices?

**Comment by dnburgess** [comment]
This looks amazing!!

**Reply by henrygd** [reply]
Thanks David, loved the video!

**Reply by dnburgess** [reply]
So glad you liked it!!

**Comment by SoNotMC** [comment]
Yes I would love this, means I can monitor all my different file systems in one place!

Also could there be a posbility of adding GPU support? Since I host a jellyfin container and being able to see GPU utilisation would be great.

**Reply by henrygd** [reply]
GPU stats are not supported by psutil or gopsutil, and apparently it's not an easy thing to add. So not likely.

GPU temp is a possibility though.

**Reply by SoNotMC** [reply]
Sounds good. nvidia-smi is a thing but im unsure on how you would get information from it due to its formatting.

**Comment by palijn** [comment]
The proposed solution looks great to me.
IF it makes sense to you (like, if the code is easier to reuse), displaying a single graph with aggregated data like the Docker usage would work pretty well for me too.

Any guess as to when you would be able to deliver this feature ? :)

**Reply by henrygd** [reply]
Hopefully by the end of the month.

**Comment by pabloalcantara** [comment]
Great. This way it´ll be best

**Comment by ctrlaltdelete007** [comment]
looks great!!!

**Comment by henrygd** [comment]
Working on implementing this now.

The `EXTRA_FILESYSTEMS` environment variable works for the binary version of the agent.

However, I can't find a way to get disk size / usage in Docker unless a folder from the device is mounted in the container.

So for the Docker agent, you would **NOT** use `EXTRA_FILESYSTEMS`, but rather do something like below. The agent would then pick up any device that has a mountpoint in `/extra-filesystems`.

```yml
volumes:
  - /mnt/disk1/.beszel:/extra-filesystems/disk1:ro
  - /mnt/disk2/.beszel:/extra-filesystems/disk2:ro
  - /dev/mmcblk0/.beszel:/extra-filesystems/sd-card:ro
```

Please let me know if this method would present issues for anyone. Or if you have an idea for a better workaround.

**Reply by JamsRepos** [reply]
I don’t see any issue with this solution, makes the most sense to me.

**Reply by crypt0rr** [reply]
Jup, all ok 👌

**Reply by svenvg93** [reply]
Yess this would work! 

Another options might be to mount /proc, /dev /sys etc into the container. I think that is how node exporter does it.  Then there is no need for mounting all the disks into the container. 

/dev: This directory contains device files, including the disks.
/sys: This directory provides information about devices, including disks, and their attributes.
/proc: This directory contains runtime system information (including information about mounted filesystems).

**Reply by JamsRepos** [reply]
@henrygd It might be worth adding the ability to change the primary file system to one of the extras too. As it seems on docker (specifically unraid) where the root system only really reports on the container itself no matter what you set, I would assume the reason being why you had to take this approach.

So it might be a good idea to change `FILESYSTEM` to also be able to be set to a volume too for the primary monitor? As otherwise you have a setup where you're collecting stats on the container which you have no care for and everything else in it's own section at the bottom of monitoring.

**Reply by henrygd** [reply]
I haven't tested, but you should be able to combine `FILESYSTEM` and `EXTRA_FILESYSTEMS` now to override which partition is picked up as root. However, you would get duplicate charts because it would still be picked up as an extra fs.

Maybe it would make sense to add a rule where you can override root by naming the mounted folder `rootfs`?

```yml
volumes:
  - /mnt/disk1/.beszel:/extra-filesystems/rootfs:ro
  # or maybe outside of /extra-filesystems is better?
  - /mnt/disk1/.beszel:/rootfs:ro
```

Edit: Actually you may be able to do `/mnt/disk1/.beszel:/rootfs:ro` currently and set `FILESYSTEM` to use that without needing to mount in `/extra-filesystems`.

**Reply by JamsRepos** [reply]
Yeah, that would make a lot more sense, as i'm sure a lot of Unraid users would want their /mnt/cache to be their /rootfs

**Reply by henrygd** [reply]
@JamsRepos I have trouble understanding the Unraid setup since I have no experience with it.

Referencing your `df` output in the Unraid discussion, it looks like your system is using `/var/lib/docker` for calculating disk usage, since that's the actual storage location for the container, but you'd prefer to use `/mnt/cache`, i.e., partition `nvme0n1p1`. So ideally it would display 30% usage instead of 19%. Do I have that right?

**Reply by JamsRepos** [reply]
@henrygd That would be correct yeah. Obviously my Unraid setup may differ from others but the ability to set this to the path of your choice would be great, although in most I would assume others would also use their cache. 

But yeah, it appears to be using `/var/lib/docker` and I'd like to change to the `nvme0n1p1` as per my command output.

**Reply by henrygd** [reply]
@JamsRepos I think this is possible now in 0.4.0 by doing this:

```yml
environment:
  FILESYSTEM: nvme0n1p1
volumes:
  - /mnt/cache/.beszel:/extra-filesystems/cache:ro
```

`FILESYSTEM` specifies which partition should be considered root, and mounting a folder gives the agent access to check the usage.

You can also try using "/extra-filesystems/cache" as the `FILESYSTEM` value.

If that doesn't work, let me know what the agent logs.

**Comment by death4444** [comment]
Technically unaware, demand +1，Looking forward to it.

**Comment by DominikBritz** [comment]
I use the beszel binary and have it running as systemd service. How do I add the environment variable `EXTRA_FILESYSTEMS ` ?

**Reply by scriptzteam** [reply]
Something like this one?

https://github.com/henrygd/beszel/blob/d2284c3fed0fa48f1b8f97902cc963f05812e098/supplemental/guides/systemd.md?plain=1#L113

```Environment="EXTRA_FILESYSTEMS=sdb"```

:)

**Reply by DominikBritz** [reply]
Perfect, thank you

**Comment by Killerbeast01** [comment]
Any posibility to change the displayed name for the disk? like i have  a lxc container with three disks, and all just get printed like "DataGrave-vm--100--disk--0", "DataGrave-vm--100--disk--1", etc...
Bind via volume mount for the docker agent:
```yaml
volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /mnt/dataGrave/.beszel:/extra-filesystems/dataGrave:ro
      - /mnt/dataGrave/Media/.beszel:/extra-filesystems/Media:ro
      - /mnt/dataGrave/Backups/.beszel:/extra-filesystems/Backups:ro
```

**Reply by henrygd** [reply]
See #379. We'll add a way to do this eventually either through a config file or maybe delimiter in the folder name.

**Reply by Justinzobel** [reply]
Was the config file option ever added? I tried it just now with the config file:
/etc/beszel-agent.conf and adding:
```
EXTRA_FILESYSTEMS="sdb,sda"
```

I then restarted using `systemctl restart beszel-agent` and checked `systemctl status beszel-agent` but saw nothing about extra filesystems/disks.

**Reply by iconoclasthero** [reply]
I believe that https://github.com/henrygd/beszel/discussions/819 is asking the same thing.

It would be nice if the displayed name was simply what was either after `... :/extra-filesystems/` or the after the final `/` such that if I mounted
```
    volumes:
      - /dir/sub1:/extra-filesystems/dir/sub1:ro
      - /dir/sub2:/estra-filesystems/dir/sub2:ro
```
it would either display `/dir/sub1`; `/dir/sub2` OR `sub`; `sub2`, respectively 

I don't know what sdb6 and sdb9 actually correspond to and they're basically identical in size and data.

**Reply by iconoclasthero** [reply]
Oh, ChatGPT said this could be done:
> Yeah — that’s actually expected behavior for Beszel Agent specifically, not Docker.
>
> Beszel doesn’t use the directory names in /extra-filesystems literally — it detects the block devices those directories live on, then reports metrics by the device name (e.g. /dev/sdb9) rather than the mountpoint or alias (/home, /library, etc.).
>
> So even though you mount:
>
> `- /library:/extra-filesystems/library:ro`
>
>
> Beszel scans /extra-filesystems/library, sees that it’s on /dev/sda, and then reports stats as /dev/sda.
>
> This is normal and intentional — Beszel uses /proc/mounts or stat device numbers to identify disks.
> 
> 👉 If you want it to show friendlier names, you can rename them in Beszel’s UI:
>
> Go to Settings → Storage,
>
> Find each detected /dev/sdX,
>
> Click the edit icon,
>
> Assign a friendly name (like “Library”, “Home”, etc.).

Which is clearly not true, but it _is_ another possibility.  It doesn't necessarily need to be done in the `docker-compose.yaml`, it _could_ be done via the PocketBase (?) settings.

**Reply by henrygd** [reply]
You should be able to do custom names now in 0.13.2 using double underscores as a separator.

#### Binary

```env
# with device name
EXTRA_FILESYSTEMS=sda1__Media
# or with mount point
EXTRA_FILESYSTEMS=/mnt/media__Media
```

#### Docker

```yaml
volumes:
  - /mnt/media/.beszel:/extra-filesystems/sda1__Media:ro
```

This will use "Media" as the name of the device (`sda1` would be used for I/O if specified).

Note that this will be picked up fresh and will not continue history from the previous name.

**Reply by Justinzobel** [reply]
I'm not getting extra filesystems loaded at all on 0.14.1 (hub and agent).

```
Oct 23 02:54:08 hostname systemd[1]: Started Beszel Agent Service.
Oct 23 02:54:08 hostname beszel-agent[415324]: 2025/10/23 02:54:08 INFO Data directory path=/var/lib/beszel-agent
Oct 23 02:54:08 hostname beszel-agent[415324]: 2025/10/23 02:54:08 INFO Detected root device name=vda1
Oct 23 02:54:08 hostname beszel-agent[415324]: 2025/10/23 02:54:08 INFO Detected network interface name=eth0 sent=3151388480 recv=39142502664
Oct 23 02:54:08 hostname beszel-agent[415324]: 2025/10/23 02:54:08 INFO Detected network interface name=eth1 sent=57918983360388 recv=5843131651296
Oct 23 02:54:09 hostname beszel-agent[415324]: 2025/10/23 02:54:09 INFO WebSocket connected host=dash.domain.com
```

```
cat /etc/beszel-agent.conf 
EXTRA_FILESYSTEMS=sda1__Data
```

**Comment by VinterSolen** [comment]
I have half-ish working extra volumes working.

Beszel-agent:
```
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /data/.beszel:/extra-filesystems/data:ro
    environment:
```

It correctly identifies storage usage of /data/ but reports 0 I/O stats, not even when i have 100's of MBps read from the /data/ drive.
/data/ is a raidz2 ZFS if that matters?

Same with docker CPU usage seems to be 0 (agent + bezsel in docker)

**Reply by henrygd** [reply]
Since it is split over multiple devices and managed by zfs, we can't automatically populate I/O.

However, you will be able to configure this in the future by telling the agent which devices make up the pool. See https://github.com/henrygd/beszel/issues/200 for updates.

For empty docker stats see [this bit of the docs](https://beszel.dev/guide/common-issues#docker-container-charts-are-empty-or-missing) or search the issues page. If it's the same as `docker stats` then we can't fix on this end. If it differs from `docker stats` then it may be a bug.
